## Task 1

First look up the coordinates of the *gyrA* gene on NCBI, these are: 1051396 to 1054146, which in python become 1051395 to 1054145 (numbering from zero).

Now read in the the super gonorrhoea mapped concensus fasta file:

In [1]:
from Bio import SeqIO
super_gc_mapped_fa = SeqIO.read('cdt-tutorial/super_gonorrhoea/super_gc_mapped.fa', 'fasta')

Extract the gene of interest and convert it to amino acid sequence:

In [2]:
super_gc_GyrA = super_gc_mapped_fa[1051395:1054146].translate()
print(super_gc_GyrA)

ID: <unknown id>
Name: <unknown name>
Description: <unknown description>
Number of features: 0
Seq('MTDATIRHDHKFALETLPVSLEDEMRKSYLDYAMSVIVGRALPDVRDGLKPVHR...EN*', HasStopCodon(ExtendedIUPACProtein(), '*'))


We can now extract the amino acid at position 91 (numbered 90 in python) in the protein:

In [3]:
super_gc_GyrA[90]

'F'

So we confirm evidence of the S91F mutation.

We now want to compare this to the reference sequence, first read in the reference, and then translate as before:

In [4]:
ref_fa = SeqIO.read('cdt-tutorial/super_gonorrhoea/reference.fa', 'fasta')
ref_GyrA = ref_fa[1051395:1054146].translate()

Now compare the two amino acid strings, e.g.:

In [5]:
for i, (aa1, aa2) in enumerate(zip(super_gc_GyrA, ref_GyrA)):
    if aa1!=aa2:
        print(i+1, aa1, aa2)

95 A G


You can see that the two sequences also differ at position 95. The wildtype position here is D, so the reference contains one mutation D95G and the super gonorrhoea another D95A

## Task 2
First let's create the BLAST database needed:

In [3]:
import os
cmd = 'makeblastdb -dbtype nucl -in cdt-tutorial/super_gonorrhoea/super_gc_contigs.fa'
os.system(cmd)

0

Check the expected output ending super_gc_contigs.fa.n*  has been generated:

In [7]:
os.listdir('cdt-tutorial/super_gonorrhoea/')

['23s_reference.fa',
 'penA_alleles.fa',
 'reference.fa',
 'super_gc_contigs.fa',
 'super_gc_mapped.fa',
 'tetM.fa',
 'super_gc_contigs.fa.ndb',
 'super_gc_contigs.fa.ntf',
 'super_gc_contigs.fa.nin',
 'super_gc_contigs.fa.nhr',
 'super_gc_contigs.fa.nsq',
 'super_gc_contigs.fa.not',
 'super_gc_contigs.fa.nto']

Now run blast using the code snippet provided:

In [4]:
from Bio.Blast import NCBIXML
from Bio.Blast.Applications import NcbiblastnCommandline
from io import StringIO

os.chdir('cdt-tutorial/super_gonorrhoea/')

geneFile = 'tetM.fa'
assemblyFile = 'super_gc_contigs.fa'

cline = NcbiblastnCommandline( query=geneFile, 
         db=assemblyFile, 
         evalue=0.01, 
         outfmt=5
         )
stdout, stderr = cline()
fp = StringIO( stdout )
blast_records = [r for r in NCBIXML.parse( fp )]

You now need to parse the output to find what you need:

In [26]:
for record in blast_records:
    print('%s has %s alignments\n'%(record.query, len(record.alignments)))
    if len(record.alignments)>0:
        hits = []
        for alignment in record.alignments:
            for hsp in alignment.hsps:
                print('tetM length: %s'%len(hsp.query))
                print('Alignment length: %s'%hsp.align_length)
                print('Number of identities: %s'%hsp.identities)
                print('Number of gaps: %s'%hsp.gaps)

tetM has 1 alignments

tetM length: 1932
Alignment length: 1932
Number of identities: 1932
Number of gaps: 0


You can see we have found a perfect match.

## Task 3

Now search for a match in the *penA* alleles:

In [27]:
geneFile = 'penA_alleles.fa'
assemblyFile = 'super_gc_contigs.fa'

cline = NcbiblastnCommandline( query=geneFile, 
         db=assemblyFile, 
         evalue=0.01, 
         outfmt=5
         )
stdout, stderr = cline()
fp = StringIO( stdout )
blast_records = [r for r in NCBIXML.parse( fp )]

If we parse the results of the blast search as before we have lots of hits

In [42]:
for record in blast_records:
    print('%s has %s alignments'%(record.query, len(record.alignments)))
    if len(record.alignments)>0:
        hits = []
        for alignment in record.alignments:
            for hsp in alignment.hsps:
                print('penA length: %s'%len(hsp.query))
                print('Alignment length: %s'%hsp.align_length)
                print('Number of identities: %s'%hsp.identities)
                print('Number of gaps: %s\n'%hsp.gaps)

penA_1.001 has 1 alignments
penA length: 1756
Alignment length: 1756
Number of identities: 1585
Number of gaps: 14

penA_1.002 has 1 alignments
penA length: 1756
Alignment length: 1756
Number of identities: 1583
Number of gaps: 14

penA_1.003 has 1 alignments
penA length: 1756
Alignment length: 1756
Number of identities: 1584
Number of gaps: 14

penA_1.004 has 1 alignments
penA length: 1756
Alignment length: 1756
Number of identities: 1584
Number of gaps: 14

penA_1.005 has 1 alignments
penA length: 1756
Alignment length: 1756
Number of identities: 1584
Number of gaps: 14

penA_1.006 has 1 alignments
penA length: 1756
Alignment length: 1756
Number of identities: 1583
Number of gaps: 14

penA_1.007 has 1 alignments
penA length: 1756
Alignment length: 1756
Number of identities: 1583
Number of gaps: 14

penA_1.008 has 1 alignments
penA length: 1756
Alignment length: 1756
Number of identities: 1585
Number of gaps: 14

penA_1.009 has 1 alignments
penA length: 1756
Alignment length: 1756
Num

We now need to find the best hit, e.g. the one with the highest number of identities:

In [48]:
hsp_list = []
for record in blast_records:
    if len(record.alignments)>0:
        hits = []
        for alignment in record.alignments:
            for hsp in alignment.hsps:
                hsp_summary = [record.query, hsp.align_length, hsp.identities, hsp.gaps, hsp.sbjct]
                hsp_list.append(hsp_summary)

sorted_hsps = sorted(hsp_list, key=lambda hsp_list: hsp_list[2], reverse=True)
print(sorted_hsps[0])

['penA_60.001', 1749, 1749, 0, 'ATGTTGATTAAAAGCGAATATAAGCCCCGGATGCTGCCCAAAGAAGAGCAGGTCAAAAAGCCGATGACCAGTAACGGACGGATTAGCTTCGTCCTGATGGCAATGGCGGTCTTGTTTGCCTGTCTGATTGCCCGCGGGCTGTATCTGCAGACGGTAACGTATAACTTTTTGAAAGAACAGGGCGACAACCGGATTGTGCGGACTCAAGCATTGCCGGCTACACGCGGTACGGTTTCGGACCGGAACGGTGCGGTTTTGGCGTTGAGCGCGCCGACGGAGTCCCTGTTTGCCGTGCCTAAAGATATGAAGGAAATGCCGTCTGCCGCCCAATTGGAACGCCTGTCCGAGCTTGTCGATGTGCCGGTCGATGTTTTGAGGAACAAACTCGAACAGAAAGGCAAGTCGTTTATTTGGATCAAGCGGCAGCTCGATCCCAAGGTTGCCGAAGAGGTCAAAGCCTTGGGTTTGGAAAACTTTGTATTTGAAAAAGAATTAAAACGCCATTACCCGATGGGCAACCTGTTTGCACACGTCATCGGATTTACCGATATTGACGGCAAAGGTCAGGAAGGTTTGGAACTTTCGCTTGAAGACAGCCTGTATGGCGAAGACGGCGCGGAAGTTGTTTTGCGGGACCGGCAGGGCAATATTGTGGACAGCTTGGACTCCCCGCGCAATAAAGCACCGCAAAACGGCAAAGACATCATCCTTTCCCTCGATCAGAGGATTCAGACCTTGGCCTATGAAGAGTTGAACAAGGCGGTCGAATACCATCAGGCAAAAGCCGGAACGGTGGTGGTTTTGGATGCCCGCACGGGGGAAATCCTCGCCTTGGCCAATACGCCCGCCTACGATCCCAACAGACCCGGCCGGGCAGACAGCGAACAGCGGCGCAACCGTGCCGTTACCGACATGATCGAACCTGGTTCTGTCATGAAGCCGTTTACCATTGCCAAAGCATTGGATTC

The top his is *penA* 60.001, and this an exact match. We now want to analyse the amino acids in this sequence at positions 311 316 and 483. First convert the sequence matched back to a Biopython sequence object

In [51]:
from Bio.SeqRecord import SeqRecord
from Bio.Seq import Seq

top_hit = sorted_hsps[0]
seq = SeqRecord(Seq(top_hit[4]), name=top_hit[0])
print(seq)

ID: <unknown id>
Name: penA_60.001
Description: <unknown description>
Number of features: 0
Seq('ATGTTGATTAAAAGCGAATATAAGCCCCGGATGCTGCCCAAAGAAGAGCAGGTC...TAA')


In [53]:
protein = seq.translate()
print(protein[310], protein[315], protein[482])

V T S


You can see that A311V and T483S are present, but T316P. This is sufficient to confer ceftriaxone resistance.

## Task 4

We need to index the 23S gene reference:

In [54]:
cmd = 'bwa index 23s_reference.fa'
os.system(cmd)

0

Check for the index files ending 23s_reference.fa.*

In [56]:
os.listdir('.')

['23s_reference.fa',
 'penA_alleles.fa',
 'reference.fa',
 'super_gc_contigs.fa',
 'super_gc_mapped.fa',
 'tetM.fa',
 'super_gc_contigs.fa.ndb',
 'super_gc_contigs.fa.ntf',
 'super_gc_contigs.fa.nin',
 'super_gc_contigs.fa.nhr',
 'super_gc_contigs.fa.nsq',
 'super_gc_contigs.fa.not',
 'super_gc_contigs.fa.nto',
 '23s_reference.fa.pac',
 '23s_reference.fa.ann',
 '23s_reference.fa.amb',
 '23s_reference.fa.bwt',
 '23s_reference.fa.sa']

We now need to map the raw sequence read files to the new reference we have created. This will take several minutes to run, wait for the output below to include a zero before proceeding...

In [58]:
ref = '23s_reference.fa'
fq1 = '/data/cdtshared/Cohort-2020/wgs_practical/ERR2560139_1.fastq.gz'
fq2 = '/data/cdtshared/Cohort-2020/wgs_practical/ERR2560139_2.fastq.gz'
out_sam = '23s_mapped.sam'
cmd = 'bwa mem -t 4 %s %s %s > %s'%(ref, fq1, fq2, out_sam)
print(cmd)
os.system(cmd)

bwa mem -t 4 23s_reference.fa /data/cdtshared/Cohort-2020/wgs_practical/ERR2560139_1.fastq.gz /data/cdtshared/Cohort-2020/wgs_practical/ERR2560139_2.fastq.gz > 23s_mapped.sam


0

Now put only the mapped reads into their own bam file and index it:

In [60]:
out_bam = '23s_mapped.bam'
cmd = 'samtools view -uS -F 4 %s | samtools sort - -o %s'%(out_sam, out_bam)
print(cmd)
os.system(cmd)

cmd = 'samtools index %s'%out_bam
print(cmd)
os.system(cmd)

samtools view -uS -F 4 23s_mapped.sam | samtools sort - -o 23s_mapped.bam
samtools index 23s_mapped.bam


0

Now use pysam to analyse the mapped bam file, here we can also filter the reads based on quality, but this is an extra that is not part of the practical as written:

In [62]:
import pysam
samfile = pysam.Samfile(out_bam, "rb")

base2045 = []
base2597 = []

for pileupcolumn in samfile.pileup( '23s_rna', 2000, 2600):
    for pileupread in pileupcolumn.pileups:
        if pileupcolumn.pos == 2044 and pileupread.query_position != None:
            base = pileupread.alignment.query_sequence[pileupread.query_position]
            quality = ord(pileupread.alignment.qual[pileupread.query_position])
            if quality-33>=30:
                base2045.append(base)
        elif pileupcolumn.pos == 2596 and pileupread.query_position != None:
            base = pileupread.alignment.query_sequence[pileupread.query_position]
            quality = ord(pileupread.alignment.qual[pileupread.query_position])
            if quality-33>=30:
                base2597.append(base)

#check for base counts at base 2045, i.e. for evidence of A2059G
a = len([b for b in base2045 if b=='A'])
c = len([b for b in base2045 if b=='C'])
g = len([b for b in base2045 if b=='G'])
t = len([b for b in base2045 if b=='T'])

print(a, c, g, t)

0 1 5021 0


We find evidence of 4 copies of the A2059G mutation as nearly all the high quality bases are G. This will confer high level resistance to azithromycin.

We could do the same for C2611T:

In [63]:
a = len([b for b in base2597 if b=='A'])
c = len([b for b in base2597 if b=='C'])
g = len([b for b in base2597 if b=='G'])
t = len([b for b in base2597 if b=='T'])

print(a, c, g, t)

3 5342 3 1


Here we see there is only evidence of the wildtype. However, we have already explained the azithromycin resistance with the 4 copies of the A2059G mutation.

## Task 5
Well done you've made it to the final task. There are lots of ways to answer this task. I'll just provide an potential example, but your approach may well be better!

Read in the data as two pandas dataframes and just look at the first few rows to get a feel for the data

In [5]:
import pandas as pd
os.chdir('../hospital_transmission/')

samples = pd.read_csv('sample_list.csv')
snps = pd.read_csv('pairwise_snps.csv')

In [6]:
samples[0:10]

collectdt site spec_num   rt                                  guid  \
0  2013-01-19    8        1   20  8300636c-ba5e-4cce-9392-c1f2a873c6df   
1  2013-03-17    8        2   78  cf3a6a18-68e6-4ec0-a2ca-7012cade9be1   
2  2013-01-14    8        3   39  85e8cf63-28ae-4d26-b639-4b8857838793   
3  2013-01-25    8        4   18  2658e7aa-768b-4975-8642-098ae9e40397   
4  2013-02-07    8        5   14  f3ad5cb3-e94f-4e97-9ee9-ebeaf62742d0   
5  2013-03-12    8        6    2  0da19dd8-ce8c-4669-ba7a-47ac62944f96   
6  2013-03-13    8        7    2  ee2c5fe7-fccf-4668-8ef4-5a8243006a0a   
7  2013-01-13    8        8    2  9e79d0bf-a55b-49ba-a138-5ebd48e52f55   
8  2013-01-17    8        9  705  3d2d3cf4-cdfb-45c2-abed-7baf20f247b6   
9  2013-01-10    8       10   15  1fc337a2-cf18-4a65-8937-34c1db315cae   

         id fecal_tox  toxigenic  
0  8300636c       NEG          1  
1  cf3a6a18       NEG          1  
2  85e8cf63       NEG          1  
3  2658e7aa       POS          1  
4  f3ad5cb3       POS          0  
5  0da19dd8       NEG          1  
6  ee2c5fe7       NEG          1  
7  9e79d0bf       NEG          1  
8  3d2d3cf4       POS          1  
9  1fc337a2       NEG          1

We want to use the fecal toxin column later setting the EQ (equivocal values) to be POS (positive) for the purposes of simplifying this analysis.

In [7]:
samples.groupby('fecal_tox').count()

collectdt  site  spec_num   rt  guid   id  toxigenic
fecal_tox                                                      
EQ                10    10        10   10    10   10         10
NEG              510   510       510  510   510  510        510
POS              453   453       453  453   453  453        453

In [8]:
samples.loc[samples['fecal_tox'] == 'EQ', 'fecal_tox'] = 'POS'
samples.groupby('fecal_tox').count()

collectdt  site  spec_num   rt  guid   id  toxigenic
fecal_tox                                                      
NEG              510   510       510  510   510  510        510
POS              463   463       463  463   463  463        463

In [9]:
snps[0:10]

id1       id2    snps
0  2658e7aa  8300636c   14438
1  0da19dd8  8300636c   11560
2  3d2d3cf4  8300636c      29
3  1fc337a2  8300636c   13376
4  34096524  8300636c   12077
5  419fff1f  8300636c   96698
6  3f534ab6  8300636c   12263
7  3007bd9d  8300636c  100244
8  17783a9f  8300636c   12872
9  2a781b87  8300636c   11147

We want to ignore the non-toxigenic cases, so need to create a new list of samples:

In [10]:
print(len(samples))
toxigenic_samples = samples[samples['toxigenic']==1]
print(len(toxigenic_samples))

973
853


You can see that we have removed 120 non-toxigen samples, leave 853 to work with.

### Define the amount of within hospital transmission
We now want to define which hospital has the most transmission. We could define this for example as the proportion of cases with another earlier case in the dataset within ≤2 SNPs.

Let's first reduce the snps table to only include entries with ≤2 SNPs:

In [12]:
print(len(snps))
snps2 = snps[snps['snps']<=2]
print(len(snps2))

472878
556


To enable matching in either direction we now need to create a duplicate of the snps2 dataframe, but with the id1 and id2 reversed:

In [13]:
snps2_rev = snps2[['id2', 'id1', 'snps']]
snps2_rev.columns = ('id1', 'id2', 'snps') #rename columns so that are reversed when joined
snps2_all = pd.concat([snps2, snps2_rev], sort=False)
len(snps2_all.drop_duplicates())

1112

Now we need to merge the shorter SNPs table with the sample data, this will end up with fewer rows as were are only including the toxigenic cases:

In [14]:
j1 = pd.merge(snps2_all, toxigenic_samples, how='inner', left_on='id1', right_on='id')
merged = pd.merge(j1, toxigenic_samples, how='inner', left_on='id2', right_on='id', suffixes=('_1', '_2'))
print(len(merged))

1012


In [15]:
merged

id1       id2  snps collectdt_1 site_1 spec_num_1  rt_1  \
0     419fff1f  cf3a6a18     1  2013-01-27      8         12    78   
1     594fa74f  88072c26     2  2013-10-10      9         16     5   
2     9e801c22  88072c26     2  2013-07-18      d         11    27   
3     b1297db0  88072c26     2  2013-09-10      9          4     5   
4     594fa74f  b1297db0     0  2013-10-10      9         16     5   
5     88072c26  b1297db0     2  2013-02-05      8         14    13   
6     9e801c22  b1297db0     0  2013-07-18      d         11    27   
7     594fa74f  9e801c22     0  2013-10-10      9         16     5   
8     88072c26  9e801c22     2  2013-02-05      8         14    13   
9     b1297db0  9e801c22     0  2013-09-10      9          4     5   
10    01686bca  3007bd9d     1  2013-10-24      9         21    12   
11    88072c26  594fa74f     2  2013-02-05      8         14    13   
12    9e801c22  594fa74f     0  2013-07-18      d         11    27   
13    b1297db0  594fa74f     0  2013-09-10      9          4     5   
14    73740cb9  9476308f     0  2013-10-28      9         22    78   
15    4b17cf00  9476308f     2  2013-11-26      1        181    78   
16    73740cb9  4b17cf00     2  2013-10-28      9         22    78   
17    369774fe  4b17cf00     2  2014-03-14      b         90    78   
18    9476308f  4b17cf00     2  2013-09-20      9          9    78   
19    4b17cf00  73740cb9     2  2013-11-26      1        181    78   
20    9476308f  73740cb9     0  2013-09-20      9          9    78   
21    4b17cf00  369774fe     2  2013-11-26      1        181    78   
22    3ff390d4  369774fe     1  2014-01-15      c         26    78   
23    3ec5feb4  369774fe     2  2013-09-04      c         55    78   
24    5f6c6463  369774fe     1  2014-03-12      c         79    78   
25    a415226e  369774fe     1  2014-09-28      1        199    78   
26    0b5eeea5  369774fe     2  2014-03-05      1         12    78   
27    5179e27a  369774fe     2  2014-06-19      1         42    78   
28    7167edcb  369774fe     2  2014-09-06      1        188    78   
29    a4664978  369774fe     2  2014-06-04      b        124    78   
...        ...       ...   ...         ...    ...        ...   ...   
982   f92505e9  3d2e6654     0  2013-10-30      b         34    14   
983   d96fbeae  6d506282     0  2013-10-31      b         36    20   
984   da96f0d2  a3ae968d     0  2013-11-04      b         37    46   
985   6f23e633  249567fd     0  2013-12-20      b         55     5   
986   e887bc65  423dc7f0     0  2014-01-16      b         68    29   
987   752ecfe0  1f220a3d     1  2014-01-22      b         73    54   
988   a2d79768  0561f5e5     0  2014-02-17      b         86    20   
989   ebb8c307  12c36113     0  2014-03-26      b         97     3   
990   e69aebfc  e295c81b     2  2014-04-22      b        106    87   
991   ef371075  15b79242     1  2014-06-02      b        121     5   
992   9509b554  4a495a33     0  2014-06-04      b        123    17   
993   f2931556  e66c2be7     0  2014-06-19      b        129    27   
994   875b85c5  84298c53     2  2014-08-31      b        147    14   
995   31e1c4b6  30b986bd     0  2013-12-07      b        153   283   
996   b38315f1  3dfa41cd     0  2014-07-06      b        161     2   
997   c2f9fe13  078163ca     1  2014-05-10      b        163     2   
998   a17b40e7  9e784496     0  2014-08-05      b        171    14   
999   cb77a6c9  4cd92de6     1  2014-08-18      b        178    78   
1000  6c53e096  653f481f     1  2014-06-12      b        183   167   
1001  b1f76a0f  6648a74f     0  2013-09-06      9        103     2   
1002  626d7760  01179179     0  2013-12-18      9        156   293   
1003  eeee1962  be738ce9     0  2013-12-25      9        164   619   
1004  83016154  08abd73d     1  2014-04-22      9        217     5   
1005  c8c6900d  3798e939     0  2014-07-02      9        251    18   
1006  e011ca6b  c2920ca5     2  2014-07-09      9        259    50   
1007  f8bb5d28  86a0e

We can now filter this down to ask how many of the pairs have an earlier case as case 1:

In [16]:
merged_ordered = merged[merged['collectdt_1']<merged['collectdt_2']]
len(merged_ordered)

489

Now restrict to only pairs in the same hospital:

In [17]:
merged_hospital = merged_ordered[merged_ordered['site_1']==merged_ordered['site_2']]
len(merged_hospital)

271

Now get a count of the number of cases at each hospital with a prior case within ≤ 2 SNPs, and compare this to the total number of cases. We need to only count one link per recipient case.

In [25]:
links = merged_hospital.groupby('site_1').id_2.nunique()
df_links = links.rename_axis('site').reset_index(name='links')
df_links

site  links
0    1     34
1    8     15
2    9     47
3    b     36
4    c     21
5    d     13

In [26]:
cases = toxigenic_samples.groupby('site').count()['id']
df_cases = cases.rename_axis('site').reset_index(name='cases')
df_cases

site  cases
0    1    175
1    8    136
2    9    183
3    b    162
4    c    110
5    d     87

In [27]:
summary = pd.merge(df_cases, df_links)
summary['pct_liinked'] = round(100*summary['links'] / summary['cases'],1)
summary

site  cases  links  pct_liinked
0    1    175     34         19.4
1    8    136     15         11.0
2    9    183     47         25.7
3    b    162     36         22.2
4    c    110     21         19.1
5    d     87     13         14.9

Site 9 has the most transmission, and site 8 the least using this metric. In the paper we only consider recipients occuring after the first 90 days of the study at each hospital, you could considered trying this here.

### Define which cases are more infectious

We will use the same dataframe merged_hospital, but group by the fecal toxin column for the source case, i.e. the earlier case, case 1, and count the number of unique second cases linked to each fecal toxin type.

In [28]:
tmp = merged_hospital.groupby('fecal_tox_1').id_2.nunique()
links_tox = tmp.rename_axis('fecal_tox').reset_index(name='links')
links_tox

fecal_tox  links
0       NEG    105
1       POS     92

In [29]:
tmp = toxigenic_samples.groupby('fecal_tox').count()['id']
cases_tox = tmp.rename_axis('fecal_tox').reset_index(name='cases')
cases_tox

fecal_tox  cases
0       NEG    405
1       POS    448

In [30]:
summary_tox = pd.merge(cases_tox, links_tox)
summary_tox['pct_liinked'] = round(100*summary_tox['links'] / summary_tox['cases'],1)
summary_tox

fecal_tox  cases  links  pct_liinked
0       NEG    405    105         25.9
1       POS    448     92         20.5

Potentially rates of linkage to a fecal toxin positive case are higher. However you could improve on this by considering the number of unique cases linked to a fecal toxin positive case only, a fecal toxin negative case only, or both. You are probably too short on time to do this today - but you could look at what we do in the paper if you have time.

### Transmission between hospitals

You know the answer is yes, from the above. But here is a table with possible links, there are 218. Is this suprising?

In [24]:
merged_ordered[merged_ordered['site_1']!=merged_ordered['site_2']]

id1       id2  snps collectdt_1 site_1 spec_num_1  rt_1  \
5     88072c26  b1297db0     2  2013-02-05      8         14    13   
6     9e801c22  b1297db0     0  2013-07-18      d         11    27   
8     88072c26  9e801c22     2  2013-02-05      8         14    13   
11    88072c26  594fa74f     2  2013-02-05      8         14    13   
12    9e801c22  594fa74f     0  2013-07-18      d         11    27   
16    73740cb9  4b17cf00     2  2013-10-28      9         22    78   
18    9476308f  4b17cf00     2  2013-09-20      9          9    78   
21    4b17cf00  369774fe     2  2013-11-26      1        181    78   
22    3ff390d4  369774fe     1  2014-01-15      c         26    78   
23    3ec5feb4  369774fe     2  2013-09-04      c         55    78   
24    5f6c6463  369774fe     1  2014-03-12      c         79    78   
26    0b5eeea5  369774fe     2  2014-03-05      1         12    78   
32    b435f72c  369774fe     1  2013-10-09      8        107    78   
34    be235a1d  369774fe     1  2013-10-29      9        126    78   
35    07ac86dd  369774fe     2  2013-03-08      8         38    27   
36    eda4a20a  369774fe     2  2013-12-03      1         88   126   
39    044441ed  6b82cd14     1  2013-03-30      8         19    23   
73    8204bc8a  710c0443     2  2013-11-30      8        129    14   
74    fee784f7  710c0443     1  2013-02-13      8         26    20   
90    33fb3aa7  a7cf0fa2     0  2013-04-13      8        137    18   
91    6c1a4402  9c06f7c5     0  2013-10-01      8         74    11   
93    d3bcc27e  5eb7d69c     0  2013-01-27      8        140    78   
97    50f00375  5a990862     2  2013-05-19      8        135    12   
98    1caf9688  e4b4dc79     0  2013-06-11      8         55    12   
141   8300636c  d81e4546     0  2013-01-19      8          1    20   
143   0971d9c9  da92d478     0  2013-10-19      c         59    87   
144   596cd535  da92d478     0  2014-01-23      c         69    87   
150   4e2f25c8  da92d478     0  2014-01-27      b         76    87   
151   48acdc8a  da92d478     0  2014-01-27      b         79    87   
152   c23368ad  da92d478     0  2014-02-05      b         81    87   
...        ...       ...   ...         ...    ...        ...   ...   
715   974182c5  909a0795     0  2014-06-20      b        159    27   
716   e691c825  909a0795     0  2014-07-11      b        140    27   
730   10e21ffa  fcb8c702     0  2014-06-19      b        130    27   
734   974182c5  fcb8c702     0  2014-06-20      b        159    27   
735   e691c825  fcb8c702     0  2014-07-11      b        140    27   
761   47d88a3d  96dfbc8a     2  2013-12-04      9        144    14   
771   7dbb273d  f4743495     2  2013-09-19      b         11    50   
772   7dbb273d  30eb4183     2  2013-09-19      b         11    50   
783   b5448736  c1755dcd     0  2014-03-19      1        194    18   
792   3d2e6654  f92505e9     0  2013-10-01      9        116    14   
795   249567fd  6f23e633     0  2013-09-30      c          8     5   
839   1f220a3d  752ecfe0     1  2013-10-08      8         75    54   
840   50b80dc5  0ff02e9d     1  2013-06-19      d          7    11   
844   57f2fb58  ef93317d     0  2013-12-23      9        161    59   
845   f01a65a1  ef93317d     0  2013-12-18      9       157a    59   
848   e295c81b  e69aebfc     2  2013-11-05      c         12    87   
849   15b79242  ef371075     1  2013-08-03      8         64   174   
882   71acaeaa  52db711f     2  2013-12-30      8        130    14   
889   b5448736  8829c770     0  2014-03-19      1        194    18   
893   08abd73d  83016154     1  2013-06-07      8         57     5   
895   c2920ca5  e011ca6b     2  2014-04-05      c         39    50   
915   3007bd9d  01686bca     1  2013-01-27      8         15    78   
918   99dcf08a  032777d4     1  2013-05-19      8         36    12   
921   c0d23f55  72b44805     0  2013-11-26      9         30    14   
949   71aa4d87  36d07fe3     0  2013-10-29      c         11    15   
962   417b0bce  24c6c